sky/packages/sky/lib/src/rendering/shifted_box.dart

```diff
@@ -205,6 +205,103 @@ class RenderPositionedBox extends RenderShiftedBox {
 }
 
+typedef Size GetSizeCallback(BoxConstraints constraints);
+typedef BoxConstraints GetInnerConstraintsCallback(BoxConstraints constraints);
+typedef Point GetChildPosition(Size size, Size childSize);
+
+class RenderCustomShiftedBox extends RenderShiftedBox {
+  RenderCustomShiftedBox({
+    RenderBox child,
+    GetSizeCallback getSize,
+    GetInnerConstraintsCallback getInnerConstraints,
+    GetChildPosition getChildPosition
+  }) : _getSize = getSize,
+       _getInnerConstraints = getInnerConstraints,
+       _getChildPosition = getChildPosition,
+       super(child);
+
+  GetSizeCallback get getSize => _getSize;
+  GetSizeCallback _getSize;
+  void set getSize (GetSizeCallback newCallback) {
+    assert(newCallback != null || !attached);
+    if (_getSize == newCallback)
+      return;
+    _getSize = newCallback;
+    markNeedsLayout();
+  }
+
+  GetInnerConstraintsCallback get getInnerConstraints => _getInnerConstraints;
+  GetInnerConstraintsCallback _getInnerConstraints;
+  void set getInnerConstraints (GetInnerConstraintsCallback newCallback) {
+    assert(newCallback != null || !attached);
+    if (_getInnerConstraints == newCallback)
+      return;
+    _getInnerConstraints = newCallback;
+    markNeedsLayout();
+  }
+
+  GetChildPosition get getChildPosition => _getChildPosition;
+  GetChildPosition _getChildPosition;
+  void set getChildPosition (GetChildPosition newCallback) {
+    assert(newCallback != null || !attached);
+    if (_getChildPosition == newCallback)
+      return;
+    _getChildPosition = newCallback;
+    markNeedsLayout();
+  }
+
+  double _getIntrinsicWidth(BoxConstraints constraints) {
+    if (_getSize != null)
+      return constraints.constrain(_getSize(constraints)).width;
+    return constraints.constrainWidth();
+  }
+
+  double getMinIntrinsicWidth(BoxConstraints constraints) {
+    return _getIntrinsicWidth(constraints);
+  }
+
+  double getMaxIntrinsicWidth(BoxConstraints constraints) {
+    return _getIntrinsicWidth(constraints);
+  }
+
+  double _getIntrinsicHeight(BoxConstraints constraints) {
+    if (_getSize != null)
+      return constraints.constrain(_getSize(constraints)).height;
+    return constraints.constrainHeight();
+  }
+
+  double getMinIntrinsicHeight(BoxConstraints constraints) {
+    return _getIntrinsicHeight(constraints);
+  }
+
+  double getMaxIntrinsicHeight(BoxConstraints constraints) {
+    return _getIntrinsicHeight(constraints);
+  }
+
+  bool get sizedByParent => true;
+
+  void performResize() {
+    if (_getSize != null)
+      size = constraints.constrain(_getSize(constraints));
+    else
+      size = constraints.biggest;
+  }
+
+  void performLayout() {
+    if (child != null) {
+      BoxConstraints innerConstraints = constraints;
+      if (_getInnerConstraints != null)
+        innerConstraints = _getInnerConstraints(constraints);
+      final bool usesSize = (_getChildPosition != null);
+      child.layout(constraints.loosen(), parentUsesSize: usesSize);
```

Comment:
Why do we take the constraints the child (typically) has reported about itself and then apply loosen()?